samples/net/google_iot_device/src/bsp/iotc_bsp_io_net_zephyr.c

```diff
@@ -198,9 +206,9 @@ iotc_bsp_io_net_state_t iotc_bsp_io_net_select(
   }
 
   /* call the actual posix select */
-  const int result = poll(fds, 1, timeout_sec);
+  const int result = poll(fds, socket_events_array_size, timeout_sec);
 
-  if (0 < result) {
+  if (socket_events_array_size == 0 || 0 < result) {
```

Comment:
I've tested on `native_posix` build but `socket_events_array_size` is never 0 and `poll` does not consider the `timeout_sec` parameter, it returns immediately. I know `poll` does not equal `select` but then why does it take the timeout parameter? My hope and assumption the issue is about lack of implementation in `native_posix`. How does it behave on qemu_x86?

Reply:
Additionally the library's intended behaviour is to always listen on the socket. In order to be responsive on incoming messages (PUBLISH, PINGREQ/PINGRESP, acks) which means there should be always at least one element in the array: the `want_read` socket event.
Did you experienced zero array size?
AFAIS there are two issues:
1) If array size is zero the function becomes meaningless, it just returns `IOTC_BSP_IO_NET_STATE_OK` without listening on incoming message or waiting for the socket to be able to send bytes.
2) If poll ignores timeout then the library will do "busy-wait" which will melt down the processor and eat up the battery.

Reply:
@atigyi 

> I've tested on native_posix build but socket_events_array_size is never 0 

I found that it is (can be) 0 during startup. I didn't have an issue with that on native_posix or with my older qemu_x86 port, but doing that now, I found that it consistently errors out during initialization. Investigating that, I found that it's called once with socket_events_array_size == 0, when completely random on-stack data is passed to poll(). The actual behavior thus may vary (again, it was 100% of cases error return and halt with qemu_x86).

> poll does not consider the timeout_sec parameter, it returns immediately.

Would need to look into that in more detail. It should wait till timeout in case events are not available, thought currently Zephyr considers a socket always writable.

All I can say, that with this patch (and with the other patches still required), the sample works as expected for me on qemu_x86.

Reply:
> 2. If poll ignores timeout then the library will do "busy-wait" which will melt down the processor and eat up the battery.

Rest assured that I'm aware of this case/implications, we're actually discussing this matter in upstream in e.g. https://github.com/zephyrproject-rtos/zephyr/pull/18879#issuecomment-528331816